### PR TITLE
KTOR-604 Add file and mailto protocols support

### LINKED ISSUE #1204: Url(…) doesn't parse non-HTTP-like URL strings correctly
`Url` is [supposed to](https://kotlinlang.slack.com/archives/C0A974TJ9/p1561553541222200?thread_ts=1561546449.221000&cid=C0A974TJ9) not just support HTTP-like URLs but [any kind of URLs](https://tools.ietf.org/html/rfc1738).

However the string parser seems to properly support only HTTP-like URLs:

### Example
```kotlin
Url("mailto:marc@knaup.io").toString()
Url("file:///var/www").toString()
```

### Expected
```kotlin
"mailto:marc@knaup.io"
"file:///var/www"
```

### Actual
```kotlin
"mailto://localhost/marc@knaup.io"
"file://var/www"
```

Ktor 1.2.2

### PR DESCRIPTION
**Pull request by e5l**
Fixes KTOR-604
Fixes #1204 

Rebased #1253

### CODE REVIEW

File: ktor-http/common/src/io/ktor/http/URLParser.kt [hunk 1]
```diff
@@ -41,6 +41,30 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
     val slashCount = count(urlString, startIndex, endIndex, '/')
     startIndex += slashCount
 
+    if (protocol.name == "file") {
+        require(slashCount == 3)
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLParser.kt:45]: A message is required: we don't support host in file URIs so we need to clearly state it

File: ktor-http/common/src/io/ktor/http/URLParser.kt [hunk 2]
```diff
@@ -41,6 +41,30 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
...
+        require(slashCount == 0)
+
+        val delimiter = urlString.indexOf("@", startIndex);
+        val passwordIndex = urlString.indexOfColonInHostPort(startIndex, delimiter)
+        if (delimiter < endIndex) {
+            if (passwordIndex != -1) {
+                user = urlString.substring(startIndex, passwordIndex)
+                password = urlString.substring(passwordIndex + 1, delimiter)
+            } else {
+                user = urlString.substring(startIndex, delimiter)
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLParser.kt:60]: So if we have no `@` character, then we get an empty username and a host without a first letter, right? I am not sure if it is allowed to have no `user@` in email, do you?

File: ktor-http/common/src/io/ktor/http/URLBuilder.kt [hunk 3]
```diff
@@ -149,19 +179,42 @@ public data class Url(
...
+            append(':')
+            append(password.encodeURLParameter())
+        }
+
+        append("@")
+    }
+
+internal val URLBuilder.userAndPassword: String
+    get() = buildString {
+        val userValue = user ?: return@buildString
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLBuilder.kt:199]: duplication

File: ktor-http/common/src/io/ktor/http/URLParser.kt [hunk 4]
```diff
@@ -41,6 +41,30 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
     val slashCount = count(urlString, startIndex, endIndex, '/')
     startIndex += slashCount
 
+    if (protocol.name == "file") {
+        require(slashCount == 3)
+        host = urlString.substring(startIndex, endIndex)
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLParser.kt:46]: Why do we say that the path is a host while the host is actually missing (empty)?

File: ktor-http/common/test/io/ktor/tests/http/UrlTest.kt [hunk 5]
```diff
@@ -186,4 +189,23 @@ class UrlTest {
             assertEquals(null, password)
         }
     }
+
+    @Test
+    fun testForFileProtocol() {
+        val expectedUrl = "file:///var/www"
+        val result = Url(expectedUrl);
+        assertEquals(result.toString(), expectedUrl);
+        assertEquals(result.protocol.name, "file");
```
Comment (cy6erGn0m) [inline ktor-http/common/test/io/ktor/tests/http/UrlTest.kt:198]: assert "var/www" == result.path?

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]

File: ktor-http/common/src/io/ktor/http/URLParser.kt (hunk 1 above)
Comment (e5l) [reply ktor-http/common/src/io/ktor/http/URLParser.kt:45]: Thanks for the notice, I will add support and tests

File: ktor-http/common/src/io/ktor/http/URLParser.kt [hunk 6]
```diff
@@ -89,34 +99,72 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
...
-        rawParameters.forEach { key, values ->
-            parameters.appendAll(key, values)
+            host = urlString.substring(startIndex, nextSlash)
+            encodedPath = urlString.substring(nextSlash, endIndex)
         }
+        3 -> {
+            host = ""
+            encodedPath = "/" + urlString.substring(startIndex, endIndex)
+        }
+    }
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLParser.kt:135]: else -> error(....)

File: ktor-http/common/src/io/ktor/http/URLParser.kt [hunk 7]
```diff
@@ -89,34 +99,72 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
...
 
-        startIndex = fragmentStart
+private fun URLBuilder.parseMailto(urlString: String, startIndex: Int, endIndex: Int) {
+    val delimiter = urlString.indexOf("@", startIndex);
+    if (delimiter == -1) {
+        throw IllegalArgumentException("Invalid mailto url: $urlString, it should contains '@'.")
     }
 
-    // Fragment
+    user = urlString.substring(startIndex, delimiter)
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLParser.kt:144]: Shouldn't we decode/de-escape user and password here?

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]

File: ktor-http/common/src/io/ktor/http/URLParser.kt (hunk 7 above)
Comment (e5l) [reply ktor-http/common/src/io/ktor/http/URLParser.kt:144]: As I get it, we store everything in escaped format(the same as for regular URLs)

File: ktor-http/common/src/io/ktor/http/URLParser.kt (hunk 6 above)
Comment (e5l) [reply ktor-http/common/src/io/ktor/http/URLParser.kt:135]: Good catch, thanks

File: ktor-http/common/src/io/ktor/http/URLParser.kt (hunk 7 above)
Comment (cy6erGn0m) [reply ktor-http/common/src/io/ktor/http/URLParser.kt:145]: Then we shouldn't escape them when formatting otherwise it breaks parse-then-encode.

Comment (e5l) [reply ktor-http/common/src/io/ktor/http/URLParser.kt]: fixed as discussed, please take a look

File: ktor-http/common/src/io/ktor/http/URLParser.kt [hunk 8]
```diff
@@ -89,34 +99,73 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
...
+        }
+        else -> throw IllegalArgumentException("Invalid file url: $urlString")
+    }
+}
 
-        startIndex = fragmentStart
+private fun URLBuilder.parseMailto(urlString: String, startIndex: Int, endIndex: Int) {
+    val delimiter = urlString.indexOf("@", startIndex);
+    if (delimiter == -1) {
+        throw IllegalArgumentException("Invalid mailto url: $urlString, it should contains '@'.")
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/URLParser.kt:142]: it should contain_

**Review by cy6erGn0m — APPROVED** [review]

File: ktor-http/common/src/io/ktor/http/URLParser.kt (hunk 8 above)
Comment (e5l) [reply ktor-http/common/src/io/ktor/http/URLParser.kt]: Fixed, thanks

**Pull request merged**